sql/ddl_internal.sql

```diff
@@ -506,15 +532,9 @@ CREATE OR REPLACE FUNCTION _timescaledb_internal.truncate_hypertable(
 $BODY$
 DECLARE
     hypertable_row _timescaledb_catalog.hypertable;
+    chunk_row _timescaledb_catalog.chunk;
 BEGIN
-
-    SELECT * INTO STRICT hypertable_row
-    FROM _timescaledb_catalog.hypertable ht
-    WHERE ht.schema_name = truncate_hypertable.schema_name
-    AND ht.table_name = truncate_hypertable.table_name;
-
-    PERFORM  _timescaledb_internal.drop_chunk(c.id, FALSE)
-    FROM _timescaledb_catalog.chunk c
-    WHERE hypertable_id = hypertable_row.id;
+    --TODO: should this cascade?
```

Comment:
I think this CASCADE stuff should be determined in this PR.

I've been trying to backtrack and figure out what the discussion regarding CASCADE behavior has been thus far. But it is not clear what the decision is and the comment here seems to indicate it is undecided.

IMHO, the only way to make truncate work the same way as regular Postgres is to actually truncate the subtables instead of removing them. That would solve the cascade problem and would work with dependent objects as expected. Not clear to me why truncate isn't actually doing truncate (I know this decisions was made before, but I have doubts it was the right one). If someone really wants to delete the tables, they should call `DROP TABLE` or `drop_chunks()`. If we really want to delete the chunks during truncate, we should do it by always cascading because otherwise the TRUNCATE command might fail, which I am not sure would be right. At the very least we should in that case have an informative message saying that there are dependent objects that has to be removed first, but this is a bit weird since that is not required with regular Postgres.

Reply:
I am ok changing this to run truncate instead of drop on the chunks. I think @olofr might have an opinion here.

Reply:
I think a TRUNCATE should ideally remove chunks as the chunks are something under the hood. I don't understand why we can't let the `TRUNCATE hypertable  [ CASCADE | RESTRICT ]` just become `DROP TABLE chunk  [ CASCADE | RESTRICT ]`  with the CASCADE option inherited for each chunk. Shouldn't that have the same dependency effects as a truncate on a normal table ?

Reply:
@olofr The problem with removing chunks is that it requires also removing dependent objects, which is not the case for a TRUNCATE on a regular table. So even if I agree with the under-the-hood argument in general, this would mean our TRUNCATE would be semantically different when dropping tables under the hood. Another difference is that we won't fire any ON TRUNCATE triggers on chunks, which means we won't support ON TRUNCATE triggers on hypertables in general (in fact, it would just silently not work, unless we modify processUtility to block adding such triggers). If we really want to remove chunks I think the correct way to deal with this is to first do TRUNCATE on all chunks, and only after that we'd DROP the chunks.

Reply:
I've implemented @olofr suggestion I think CASCADE is semantically similar in TRUNCATE and DROP (in both cases dependent objects are dropped). TRUNCATE triggers can only be fired on a table level and so if you add a truncate trigger to a hypertable everything would work fine. I don't think TRUNCATE triggers on chunks make sense, and would prefer to block them.

Reply:
Namely, this "which is not the case for a TRUNCATE on a regular table" is wrong. Truncate CASCADE does remove dependent objects.

Reply:
I am not sure I agree that TRUNCATE on chunks doesn't make sense. We've talked about supporting other chunk-level operations (and customer's have asked for it afaik), so I see no reason to restricting ourselves unnecessarily in this particular case. A user might for example want to get notified in a trigger when a chunk is TRUNCATED, either to block it on a particular chunk with valuable data, or, for just simply knowing when the data covering that particular time period is deleted (due to a retention policy or for other reasons). I think we should also keep an open mind for potential future use cases we haven't thought about. 

In general, I don't think we should restrict functionality or change semantics unless (1) it is too cumbersome to implement technically, or, (2) we cannot make it work correctly.

Btw, I don't think TRUNCATE CASCADE removes, e.g., a view on a chunk since the table remains, whereas DROP will remove the view. So, implementing TRUNCATE with DROP would be different. Still, I could accept removing views since it is a bit of a corner-case, but I still think we should support TRUNCATE triggers on chunks if we can, for example by doing what I suggested: recursing TRUNCATE to chunks first, then dropping tables. That seems to be the best of both worlds and shouldn't be hard to implement.

Reply:
Will do.

Reply:
Fixed now